packages/core/src/suite.ts

```diff
@@ -1,34 +1,193 @@
+import { HookName, Hooks } from './hooks';
 import Result, { Status } from './result';
-import Runnable from './runnable';
+import Runnable, { isRunnable, RunnableOptions, RunnableTypes } from './runnable';
+import { normalizeRunOptions, RunOptions } from './runner';
 
+export const isSuite = (v: unknown): v is Suite => {
+  if (!isRunnable(v)) { return false; }
+  return v.type === RunnableTypes.Suite;
+};
+export const rootSymbol = Symbol('isRoot');
+
+export interface Stats {
+  total: number;
+  pending: number;
+  running: number;
+  done: number;
+  passed: number;
+  failed: number;
+  skipped: number;
+  todo: number;
+  time: number;
+}
+
+export class BailError extends Error {
+  constructor(message: string) {
+    super(message) /* istanbul ignore next */;
+    this.name = 'BailError';
+  }
+}
+
+/* tslint:disable:max-classes-per-file */
 export default class Suite extends Runnable {
   public children: Runnable[];
+  public [rootSymbol]?: boolean;
+  public type = RunnableTypes.Suite;
+  public options: RunnableOptions;
+  public hooks: Hooks;
+  private failed: number;
 
-  constructor(description: string, skip = false) {
-    super(description, skip);
+  /* istanbul ignore next */
+  constructor(description: string, options: Partial<RunnableOptions> = {}, parent: Suite | null) {
+    super(description, options, parent);
+    this.options = {
+      ...Runnable.normalizeOptions(options),
+    };
     this.children = [];
+    this.failed = 0;
+
+    this.hooks = {
+      afterAll: [],
+      afterEach: [],
+      beforeAll: [],
+      beforeEach: [],
+    };
   }
 
-  public addChild(child: Runnable) {
-    child.parent = this;
-    this.children.push(child);
+  public async invokeAsyncHook(name: HookName) {
+    const hook = this.hooks[name];
+    for (const fn of hook) {
+      try {
+        await fn();
+      } catch (err) {
+        console.error(`Error in ${name} hook: ${err}`);
+      }
+    }
   }
 
   /**
-   * Run a `Suite` instance return `Runnable` status:
+   * @description Add one or more `Runnable` instances to the `children` array.
+   * @param {Runnable[]} ...children
+   * @returns {Void}
+   */
+  public addChildren(...children: Runnable[]): void {
+    for (const child of children) {
+      child.parent = this;
+    }
+    this.children.push(...children);
+  }
+
+  /**
+   * @description Check that `Suite` is the root suite
    * @public
-   * @return {Result}
+   * @return {boolean}
    */
-  public run(): Result {
-    let newStatus = Status.Passed;
+  public isRoot(): boolean {
+    return Boolean(this[rootSymbol]);
+  }
+
+  /**
+   * @description Runs a `Suite` instance asynchronously returning a `Result`:
```

Comment:
No sync variant is supported
```suggestion
   * @description Runs a `Suite` instance returning a `Result`:
```